core/state/statedb.go

```diff
@@ -923,13 +863,23 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	// If the self-destruct is handled first, then `P` would be left with only one child, thus collapsed
 	// into a shortnode. This requires `B` to be resolved from disk.
 	// Whereas if the created node is handled first, then the collapse is avoided, and `B` is not resolved.
-	var deletedAddrs []common.Address
-	for addr := range s.stateObjectsPending {
-		if obj := s.stateObjects[addr]; !obj.deleted {
+	var (
+		usedAddrs    [][]byte
+		deletedAddrs []common.Address
+	)
+	for addr, obj := range s.stateObjectsPhase {
```

Comment:
Please change `obj` here to e.g. `mutation`, or `info`. It's a meta-object holding info about the mutation. As it is, it clashes with `obj` on L879, where `obj` is an actual state object. 
Let's keep them separate, making it easier to read